tests/vsa/compare_vsa.py

```diff
@@ -0,0 +1,265 @@
+# Copyright (c) 2024 - 2024, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""Script to compare a generated VSA with an expected payload."""
+
+from __future__ import annotations
+
+import argparse
+import base64
+import json
+import os
+import sys
+import traceback
+from collections.abc import Callable
+from difflib import context_diff
+from functools import partial
+from pprint import pformat
+
+# Works similarly to print, but prints to stderr by default.
+log = partial(print, file=sys.stderr)
+
+
+def log_with_tag(tag: str) -> Callable[[str], None]:
+    """Generate a log function that prints the name of the file and a tag at the beginning of each line."""
+
+    def log_fn(msg: str) -> None:
+        log(f"[{os.path.basename(__file__)} {tag}] {msg}")
+
+    return log_fn
+
+
+log_info = log_with_tag("INFO")
+log_err = log_with_tag("ERROR")
+log_failed = log_with_tag("FAILED")
+log_passed = log_with_tag("PASSED")
+
+
+def log_diff(result: object, expected: object) -> None:
+    """Pretty-print the diff of two Python objects."""
+    log(
+        "".join(
+            context_diff(pformat(result), pformat(expected), "result", "expected"),
+        ),
+    )
```

Comment:
Here, I define new log functions instead of using the built-in `logging` module for better control.

* I define some additional custom log tags (`PASSED` and `FAILED`) to indicate the compare result. These tags are not available by default in the `logging` module.
* This also allows the use of f-string without giving lint errors.

The choice of `stderr` is mostly due to convention: the output of this script is for debugging purposes mainly.

Reply:
I prefer to use the standard `logging` library. You can achieve the points you mentioned with `logging` too:

```python
logger: logging.Logger = logging.getLogger(__name__)
logging.basicConfig(format="[%(filename)s %(tag)s] %(message)s")

logger.info("The payload of the generated VSA matches the expected payload", extra={"tag": "PASSED"})
```

Will log

```
[compare_vsa.py PASSED] The payload of the generated VSA matches the expected payload.
```

I also prefer to add the line info
```
logging.basicConfig(format="[%(filename)s:%(lineno)s %(tag)s] %(message)s")
```
logs
```
[compare_vsa.py:284 PASSED] The payload of the generated VSA matches the expected payload.
```

Reply:
Thanks for the suggestion. See my change in 52c4704d5542cebd35fa3a6de9709781d529e938.